modules/extensions/src/collide/collide-effect.ts

```diff
@@ -0,0 +1,254 @@
+import {Texture2D, cssToDeviceRatio} from '@luma.gl/core';
+import {readPixelsToArray} from '@luma.gl/core';
+import type {Effect, Layer, PreRenderOptions, Viewport} from '@deck.gl/core';
+import CollidePass from './collide-pass';
+
+type CollideExtensionProps = {
+  collideTestProps?: {};
+  collideGroup: string;
+};
+
+// Factor by which to downscale Collide FBO relative to canvas
+const DOWNSCALE = 2;
+
+type RenderInfo = {
+  collideGroup: string;
+  layers: Layer<CollideExtensionProps>[];
+  layerBounds: ([number[], number[]] | null)[];
+};
+
+// Class to manage collide effect
+export default class CollideEffect implements Effect {
+  id = 'collide-effect';
+  props = null;
+  useInPicking = true;
+
+  private channels: Record<string, RenderInfo> = {};
+  private collidePasses: Record<string, CollidePass> = {};
+  private dummyCollideMap?: Texture2D;
+  private lastViewport?: Viewport;
+
+  preRender(
+    gl: WebGLRenderingContext,
+    {effects: allEffects, layers, layerFilter, viewports, onViewportActive, views}: PreRenderOptions
+  ): void {
+    if (!this.dummyCollideMap) {
+      this.dummyCollideMap = new Texture2D(gl, {width: 1, height: 1});
+    }
+
+    const collideLayers = layers.filter(
+      // @ts-ignore
+      ({props: {visible, extensions, collideGroup}}) =>
+        visible &&
+        extensions.find(e => e.constructor.extensionName === 'CollideExtension') &&
+        collideGroup
+    ) as Layer<CollideExtensionProps>[];
+    if (collideLayers.length === 0) {
+      this.channels = {};
+      return;
+    }
+
+    const effects = allEffects?.filter(({useInCollide}) => useInCollide);
+    const otherEffectRendered = Boolean(
+      effects && effects.filter(({didRender}) => didRender).length
+    );
+
+    // Collect layers to render
+    const channels = this._groupByCollideGroup(gl, collideLayers);
+
+    const viewport = viewports[0];
+    const viewportChanged =
+      !this.lastViewport || !this.lastViewport.equals(viewport) || otherEffectRendered;
+
+    // Resize framebuffers to match canvas
+    for (const collideGroup in channels) {
+      const collidePass = this.collidePasses[collideGroup];
```

Comment:
Nit: I know I set a bad precedence in `MaskPass`, but a pass shouldn't really own any resources...